bpf/bpf_overlay.c

```diff
@@ -353,7 +359,8 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 		}
 #endif
 		/* See comment at equivalent code in handle_ipv6() */
-		if (info && identity_is_remote_node(*identity))
+		if (info && (identity_is_remote_node(*identity) ||
+			     (is_dsr && identity_is_world_ipv4(*identity))))
```

Comment:
Looks good, ty! I wonder if we truly need the `identity_is_world_ipv4()` check, or should just override the source identity. But that's not a blocking change.

Have you double-checked whether we now get the same behaviour as when using "IP opt/ext" DSR? It derives the identity via https://github.com/cilium/cilium/blob/8d7dd285d75a78ab32153fbc8812de5082f721f9/bpf/bpf_host.c#L474-L476, with `srcid_from_proxy == 0` and `from_host == false`.

Reply:
> I wonder if we truly need the identity_is_world_ipv4() check, or should just always override the source identity.

I put `identity_is_world`, but as you wrote, I too think it will work well whether we check its identity or not if the packet has DSR info.
@ysksuzuki wdyt?

> Have you double-checked whether we now get the same behaviour as when using "IP opt/ext" DSR?

I've checked with my k8s, and it almost works well :)
But the identity is still `reserved:world-ipv4/6` at from-overlay when the DSR with Geneve is enabled🤔
`hubble observe` outputs are these↓

**IP opt**
```console
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) from-network FORWARDED (TCP Flags: SYN)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) policy-verdict:L3-Only INGRESS ALLOWED (TCP Flags: SYN)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: SYN)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK, PSH)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7 (ID:58676) pre-xlate-rev TRACED (TCP)
Nov 20 04:32:48.556: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7 (ID:58676) pre-xlate-rev TRACED (TCP)
Nov 20 04:32:48.578: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK)
Nov 20 04:32:48.578: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 04:32:48.578: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK, FIN)
Nov 20 04:32:48.578: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Nov 20 04:32:48.579: 172.19.0.3:59036 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK)
Nov 20 04:32:48.579: 172.19.0.3:59036 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-tcnm7:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK)
```

**IP ext**
```console
Nov 20 04:30:48.839: fc00:c111::3 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj (ID:58676) from-network FORWARDED (IPv6)
Nov 20 04:30:48.839: fc00:c111::3 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj (ID:58676) policy-verdict:L3-Only INGRESS ALLOWED (IPv6)
Nov 20 04:30:48.839: fc00:c111::3 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj (ID:58676) to-endpoint FORWARDED (IPv6)
Nov 20 04:30:48.839: [fc00:c111::3]:44532 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK)
Nov 20 04:30:48.839: [fc00:c111::3]:44532 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 04:30:48.839: [fc00:c111::3]:44532 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK, PSH)
Nov 20 04:30:48.839: [fc00:c111::3]:44532 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Nov 20 04:30:48.840: [fc00:c111::3]:44532 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj (ID:58676) pre-xlate-rev TRACED (TCP)
Nov 20 04:30:48.866: [fc00:c111::3]:44532 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK)
Nov 20 04:30:48.866: [fc00:c111::3]:44532 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 04:30:48.866: [fc00:c111::3]:44532 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK, FIN)
Nov 20 04:30:48.866: [fc00:c111::3]:44532 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Nov 20 04:30:48.869: [fc00:c111::3]:44532 (ID:16777221) <> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) from-network FORWARDED (TCP Flags: ACK)
Nov 20 04:30:48.869: [fc00:c111::3]:44532 (ID:16777221) -> test/dsr-geneve-test-67fb7f94bf-r9snj:8000 (ID:58676) to-endpoint FORWARDED (TCP Flags: ACK)
```

**geneve(ipv4)**
```console
Nov 20 07:22:07.986: 172.19.0.5:38448 (ID:9) <> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: SYN)
Nov 20 07:22:07.986: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) policy-verdict:L3-Only INGRESS ALLOWED (TCP Flags: SYN)
Nov 20 07:22:07.986: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: SYN)
Nov 20 07:22:07.987: 172.19.0.5:38448 (ID:9) <> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK)
Nov 20 07:22:07.987: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 07:22:07.988: 172.19.0.5:38448 (ID:16777219) <> test/dsr-geneve-test-67fb7f94bf-466f7 (ID:8532) pre-xlate-rev TRACED (TCP)
Nov 20 07:22:07.988: 172.19.0.5:38448 (ID:16777219) <> test/dsr-geneve-test-67fb7f94bf-466f7 (ID:8532) pre-xlate-rev TRACED (TCP)
Nov 20 07:22:07.988: 172.19.0.5:38448 (ID:9) <> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK, PSH)
Nov 20 07:22:07.988: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Nov 20 07:22:07.995: 172.19.0.5:38448 (ID:9) <> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK)
Nov 20 07:22:07.995: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 07:22:07.995: 172.19.0.5:38448 (ID:9) <> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK, FIN)
Nov 20 07:22:07.995: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Nov 20 07:22:07.996: 172.19.0.5:38448 (ID:9) <> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK)
Nov 20 07:22:07.996: 172.19.0.5:38448 (ID:16777219) -> test/dsr-geneve-test-67fb7f94bf-466f7:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK)
```

**geneve(ipv6)**
```console
Nov 20 07:34:05.059: [fc00:c111::5]:40822 (ID:10) <> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: SYN)
Nov 20 07:34:05.059: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) policy-verdict:L3-Only INGRESS ALLOWED (TCP Flags: SYN)
Nov 20 07:34:05.059: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: SYN)
Nov 20 07:34:05.060: [fc00:c111::5]:40822 (ID:10) <> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK)
Nov 20 07:34:05.060: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 07:34:05.060: [fc00:c111::5]:40822 (ID:10) <> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK, PSH)
Nov 20 07:34:05.060: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Nov 20 07:34:05.060: [fc00:c111::5]:40822 (ID:16777220) <> test/dsr-geneve-test-67fb7f94bf-dftzf (ID:8532) pre-xlate-rev TRACED (TCP)
Nov 20 07:34:05.064: [fc00:c111::5]:40822 (ID:10) <> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK, FIN)
Nov 20 07:34:05.064: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Nov 20 07:34:05.064: [fc00:c111::5]:40822 (ID:10) <> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK)
Nov 20 07:34:05.064: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK)
Nov 20 07:34:05.066: [fc00:c111::5]:40822 (ID:10) <> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) from-overlay FORWARDED (TCP Flags: ACK)
Nov 20 07:34:05.066: [fc00:c111::5]:40822 (ID:16777220) -> test/dsr-geneve-test-67fb7f94bf-dftzf:8000 (ID:8532) to-endpoint FORWARDED (TCP Flags: ACK)

```

Reply:
To fix this, we have to set identity [here](https://github.com/cilium/cilium/blob/8e4c0a2fe66684bbd8082c4147aaef0f4e73b7a6/bpf/bpf_overlay.c#L616-L618)

Reply:
> To fix this, we have to set identity here.
> 
> https://github.com/cilium/cilium/blob/8e4c0a2fe66684bbd8082c4147aaef0f4e73b7a6/bpf/bpf_overlay.c#L616-L618

Ah true. Let's tackle that in a follow-up PR? :)

Reply:
> > Have you double-checked whether we now get the same behaviour as when using "IP opt/ext" DSR?
> 
> I've checked with my k8s, and it almost works well.

I had a closer look at all the special handling in `resolve_srcid_ipv4()`, and think we're good without duplicating it.

Also realized that there is a "workaround" for this bug - enable per-EP routing, and disable BPF Host Routing. Then `to-container` will resolve the identity [on its own](https://github.com/cybozu-go/cilium/blob/d25d328955273c55814729548cb7dc32bce51c14/bpf/bpf_lxc.c#L2111).

Reply:
> Let's tackle that in a follow-up PR? :)

Yes, let's do that :D